### PR TITLE
[Performance] Leverage hashmap to accelerate CSRSliceMatrix<kDGLCUDA, IdType>

### PR DESCRIPTION
**Pull request by gpzlx1**
## Description

The old `CSRSliceMatrix<kDGLCUDA, IdType>` uses BinarySearch for cols lookup with `O(E * log(#cols))` time complexity. However, `CSRSliceMatrix<kDGLCUDA, IdType>` has a severe performance drop when the `#cols` are very large. In this PR, a hashmap is leveraged instead of BinarySearch to accelerate `CSRSliceMatrix<kDGLCUDA, IdType>`, reducing the time complexity to `O(E)`. The performance of `CSRSliceMatrix<kDGLCUDA, IdType>` is improved by 1.2-49x with an acceptable memory overhead.

## Benchmark
```python
# code
import torch
import dgl
import time
import numpy as np
from dgl.data import RedditDataset


def bench(g, seeds, iter):
    rec_list = []
    for _ in range(iter):
        tic = time.time()
        dgl.node_subgraph(g, seeds)
        torch.cuda.synchronize()
        toc = time.time()
        rec_list.append(toc - tic)
    print("Time Cost [ms]: ", np.mean(rec_list[3:]) * 1000)


if __name__ == "__main__":
    g = RedditDataset(self_loop=True)[0].to('cuda')
    selected_size = [5000, 10000, 100000, 200000]
    for s in selected_size:
        seeds = torch.arange(s).cuda().long()
        bench(g, seeds, 10)
```

**Results:**
| #seeds             | 5000 | 10000 | 100000 | 200000  |
| ------------------ | ---- | ----- | ------ | ------- |
| Old Time Cost [ms] | 0.74 | 1.26  | 108.89 | 3015.49 |
| New Time Cost [ms] | 0.60 | 0.75  | 10.37  | 62.24   |
| SpeedUp            | 1.2  | 1.7   | 10.5   | 48.5    |

## Memory Overhead
For performance, the load factor of the hashmap is in (0.25, 0.5) and the additional memory overhead is `O(#cols)`. Because `#cols` is usually less than 1 Million on GPU, the memory overhead is not significant (less than 31MB) even at a low load factor.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by yaox12** [comment]
@dgl-bot

**Comment by yaox12** [comment]
@dgl-bot

File: src/array/cuda/spmat_op_impl_csr.cu [hunk 1]
```diff
@@ -423,26 +426,130 @@ template std::vector<NDArray> CSRGetDataAndIndices<kDGLCUDA, int64_t>(
...
-      cuda::AtomicAdd(count + rpos, IdType(1));
+    IdType* hashmap_buffer, int64_t buffer_size, IdType* mask, IdType* count) {
+  assert(blockDim.x == WARP_SIZE);
+  assert(blockDim.y == BLOCK_WARPS);
+
+  int warp_id = threadIdx.y;
+  int laneid = threadIdx.x;
+  IdType out_row = blockIdx.x * TILE_SIZE + threadIdx.y;
+  IdType last_row =
+      MIN(static_cast<IdType>(blockIdx.x + 1) * TILE_SIZE, num_rows);
```
Comment (yaox12) [inline src/array/cuda/spmat_op_impl_csr.cu:531]: There is a `min()` included by `cuda_runtime.h`. Can we use that?

File: src/array/cuda/spmat_op_impl_csr.cu [hunk 2]
```diff
@@ -507,10 +608,16 @@ CSRMatrix CSRSliceMatrix(
   }
 
   // Execute SegmentMaskColKernel
-  int nb = (nnz_csr + nt - 1) / nt;
+  const int64_t num_rows = csr.num_rows;
+  constexpr int WARP_SIZE = 32;
+  constexpr int BLOCK_WARP = 256 / WARP_SIZE;
+  constexpr int TILE_SIZE = 16;
```
Comment (yaox12) [inline src/array/cuda/spmat_op_impl_csr.cu:614]: How did you choose the value of `TILE_SIZE`?

**Review by yaox12 — COMMENTED** [review]
Thanks for your contribution. I'd like to invite @ayasar70 for review since he has worked on improving `_SegmentMaskColKernel`.

File: src/array/cuda/spmat_op_impl_csr.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (gpzlx1) [reply src/array/cuda/spmat_op_impl_csr.cu]: Yes. Done.

File: src/array/cuda/spmat_op_impl_csr.cu [hunk 3]
```diff
@@ -423,26 +426,130 @@ template std::vector<NDArray> CSRGetDataAndIndices<kDGLCUDA, int64_t>(
...
+}
+
+/**
+ * @brief A hashmap designed for CSRSliceMatrix, similar in function to set,
+ * Insert api inserts an element, Query api queries if an element is in the
+ * hashmap. For performance, it can only be created and called in the cuda
+ * kernel.
  */
 template <typename IdType>
+struct NodeQueryHashmap {
```
[the remark this answers is not in the document]
Comment (gpzlx1) [reply src/array/cuda/spmat_op_impl_csr.cu:466]: In short, NodeQueryHashmap modifies the hash function and removes unnecessary memory space compared to DGL's hashmap. DGL's hashmap is designed for graph relabel and does not support element query if the element does not exist in the hashable. NodeQueryHashmap overcomes this problem.

File: src/array/cuda/spmat_op_impl_csr.cu (hunk 2 above)
Comment (gpzlx1) [reply src/array/cuda/spmat_op_impl_csr.cu:613]: When TILE_SIZE = 16, I get the best performance with my device.

File: src/array/cuda/spmat_op_impl_csr.cu [hunk 4]
```diff
@@ -423,26 +424,131 @@ template std::vector<NDArray> CSRGetDataAndIndices<kDGLCUDA, int64_t>(
...
+  key = key ^ (key >> 12);
+  key = key + (key << 2);
+  key = key ^ (key >> 4);
+  key = key * 2057;
+  key = key ^ (key >> 16);
+  return key;
+}
+
+/**
+ * @brief Thomas Wang's 64 bit Mix Function
```
Comment (frozenbugs) [inline src/array/cuda/spmat_op_impl_csr.cu:446]: ditto.

File: src/array/cuda/spmat_op_impl_csr.cu [hunk 5]
```diff
@@ -423,26 +424,131 @@ template std::vector<NDArray> CSRGetDataAndIndices<kDGLCUDA, int64_t>(
...
+  key = key ^ (key >> 14);
+  key = (key + (key << 2)) + (key << 4);
+  key = key ^ (key >> 28);
+  key = key + (key << 31);
+  return key;
+}
+
+/**
+ * @brief A hashmap designed for CSRSliceMatrix, similar in function to set,
+ * Insert api inserts an element, Query api queries if an element is in the
```
Comment (frozenbugs) [inline src/array/cuda/spmat_op_impl_csr.cu:461]: \`Insert\` to to refer to the code in the comment.

File: src/array/cuda/spmat_op_impl_csr.cu (hunk 2 above)
Comment (frozenbugs) [reply src/array/cuda/spmat_op_impl_csr.cu:613]: Add a detailed comment about those magic numbers. If needed put evidence in a google doc, and link the doc to the code.

**Comment by frozenbugs** [comment]
Do you foresee any risk could be introduced by this PR?

**Comment by gpzlx1** [comment]
@frozenbugs Theoretically, no risk is introduced. I have only modified the _SegmentMaskColKernel and the input and output of this kernel remains the same as the original version.

**Comment by yaox12** [comment]
@dgl-bot

**Comment by yaox12** [comment]
> @frozenbugs Can I use thrust in DGL?

Sure. Would it be better to use `thrust::cuda::par_nosync` as follows?
https://github.com/dmlc/dgl/blob/bf264d00fe1039eb588165aee6e414ef6ba3a08a/src/runtime/cuda/cuda_common.h#L24-L29

Since we launch thrust and _SegmentMaskColKernel on the same stream, the async policy should be safe.

**Comment by gpzlx1** [comment]
> > @frozenbugs Can I use thrust in DGL?
> 
> Sure. Would it be better to use `thrust::cuda::par_nosync` as follows?
> 
> https://github.com/dmlc/dgl/blob/bf264d00fe1039eb588165aee6e414ef6ba3a08a/src/runtime/cuda/cuda_common.h#L24-L29
> 
> Since we launch thrust and _SegmentMaskColKernel on the same stream, the async policy should be safe.

Done.

**Comment by yaox12** [comment]
@dgl-bot

**Comment by yaox12** [comment]
LGTM. I got 1.2x - 142x speedup on A100 without obvious GPU memory overheads (both cost ~3677MB). It's impressive.

**Review by yaox12 — APPROVED** [review]

**Pull request merged**